ls1intum/Artemis
# `Development`: Refactor programming exercise grading service

### Checklist
#### General


#### Server
#### Changes affecting Programming Exercises
- The changes are fully independent of the Git/CI-system.

### Motivation and Context
The method to calculate the programming exercise scores was overly complex. 

### Description
This PR refactors this method into smaller ones so that the overall score calculation process is easier to understand.
Two other complex methods in this service have been refactored as well (`processNewProgrammingExerciseResult` and `addFeedbackToStatistics`).

### Steps for Testing
The refactoring in the score calculation also fixed a small bug with test case weights all zero. Therefore, this should be manually tested.
The refactoring of the other two methods did not change any behaviour.

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise **with manual assessment enabled**

1. Grading page:
    - Set the test case weight of all test cases to zero.
    - Re-evaluate all results.
2. Look at the solution in the editor. It should have a score of 100%.
3. Change the solution such that a test case fails. The score should now be `(n-1)/n` for `n` total test cases.
4. Participate as student in the exercise. The score should always remain zero, even if test cases pass (because their weight is zero).

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage
| File | Branch | Line |
| --- | --- | --- |
| `ProgrammingExerciseGradingService` | 88% | 96% |
